src/api/habits/adapters/json.ts

```diff
@@ -26,7 +30,75 @@ const jsonHabitServiceFactory = (): HabitService => {
         }))
       );
 
-  return { addHabit, getHabits };
+  const addEntry = (
+    goals: GoalWithHabitHistory[],
+    habitId: string,
+    date: Date = new Date(),
+    quantity: number = 1
+  ) => {
+    try {
+      const updatedGoals = goals.map((goal) => ({
+        ...goal,
+        habits: goal.habits.reduce((habitsAcc, currentHabit) => {
+          let habit = currentHabit;
+
+          if (habit.id === habitId) {
+            // Add an entry here
+            const newEntries = habit.entries.concat([
+              { completionDate: date, quantity },
+            ]);
+            habit = {
+              ...habit,
+              entries: newEntries,
+            };
+          }
+
+          // Otherwise just return this habit as-is
+          return habitsAcc.concat(habit);
+        }, [] as HabitWithHistory[]),
```

Comment:
I see what you're trying to do - fetch the habit, update the entries within the habit.

You might find that it'd be clearer to do something like that instead, and more performant. But good work with getting the reducer to work - I'd suggest using the `[...habitsAcc, habit]` syntax over `concat` purely for visual instant readability. It also looks like you could use a `.map` here instead - it's more or less the same effect as the reducer. But do see if you can make the code look more like "retrieve the habit" -> "add an entry", so that we don't need to iterate through all the data.

Reply:
See my Slack comment. Essentially, a reducer and a pure immutability approach here is making it harder than expected. You likely want to do something like:
1) [optional] deep clone the object with something like `structuredClone(value)`. PS, hardly _anyone_ knows about this little gem of a method
2) Find the habit you'd like to update
3) Update the values of the habit/add the new entry
4) Save this all together

```
...
const newGoals = structuredClone(goals)

const habit = newGoals.habits.find(habit => habit.id === habitId)

if (!habit) return // or throw error here

habit.entries.push({ completionDate: date, quantity })

saveData(newGoals)
```

A great example where mutability just makes this clearer - but you want to scope this to the function so that it does not leak outside or potentially modify parameters

Reply:
`structuredClone()` doesn't work with Jest at the moment...
> ReferenceError: structuredClone is not defined

I used lodash's `cloneDeep` method instead.